spark/src/main/scala/org/apache/spark/sql/delta/commands/UpdateCommand.scala

```diff
@@ -117,76 +118,153 @@ case class UpdateCommand(
+          val dvRewriteStartMs = System.nanoTime()
+          val newFiles = rewriteFiles(
+            sparkSession,
+            txn,
+            tahoeFileIndex.path,
+            touchedFiles.map(_.fileLogEntry),
+            nameToAddFile,
+            updateCondition,
+            generateRemoveFileActions = false,
+            copyUnmodifiedRows = false)
+          rewriteTimeMs = (System.nanoTime() - dvRewriteStartMs) / 1000 / 1000
+
+          dvActions ++ newFiles
+        } else {
+          Nil // Nothing to update
+        }
+      } else {
+        // Case 3.2: Update without persistent deletion vectors. Existing DVs will be materialized
+        // Keep everything from the resolved target except a new TahoeFileIndex
+        // that only involves the affected files instead of all files.
+        val newTarget = DeltaTableUtils.replaceFileIndex(target, fileIndex)
+        val data = Dataset.ofRows(sparkSession, newTarget)
+        val incrUpdatedCountExpr = IncrementMetric(TrueLiteral, metrics("numUpdatedRows"))
+        val pathsToRewrite =
+          withStatusCode("DELTA", UpdateCommand.FINDING_TOUCHED_FILES_MSG) {
+            data.filter(new Column(updateCondition))
+              .select(input_file_name())
+              .filter(new Column(incrUpdatedCountExpr))
+              .distinct()
+              .as[String]
+              .collect()
+          }
+
+        scanTimeMs = (System.nanoTime() - startTime) / 1000 / 1000
+
+        val touchedFiles =
+          pathsToRewrite.map(getTouchedFile(deltaLog.dataPath, _, nameToAddFile)).toSeq
+
+        // Rewrite all touchedFiles
+        val addAndRemoveActions =
+          withStatusCode("DELTA", UpdateCommand.rewritingFilesMsg(touchedFiles.size)) {
+            if (touchedFiles.nonEmpty) {
+              rewriteFiles(
+                sparkSession,
+                txn,
+                tahoeFileIndex.path,
+                touchedFiles,
+                nameToAddFile,
+                updateCondition,
+                generateRemoveFileActions = true,
+                copyUnmodifiedRows = true)
+            } else {
+              Nil
+            }
+          }
+        rewriteTimeMs = (System.nanoTime() - startTime) / 1000 / 1000 - scanTimeMs
+        numTouchedFiles = touchedFiles.length
+        val (addActions, removeActions) = addAndRemoveActions.partition(_.isInstanceOf[AddFile])
+        numRewrittenFiles = addActions.size
+        numAddedBytes = addActions.map(_.getFileSize).sum
+        numRemovedBytes = removeActions.map(_.getFileSize).sum
+        addAndRemoveActions
       }
     }
 
-    rewriteTimeMs = (System.nanoTime() - startTime) / 1000 / 1000 - scanTimeMs
-
-    val (changeActions, addActions) = newActions.partition(_.isInstanceOf[AddCDCFile])
-    numRewrittenFiles = addActions.size
-    numAddedBytes = addActions.map(_.getFileSize).sum
+    val (changeActions, _) = allActions.partition(_.isInstanceOf[AddCDCFile])
```

Comment:
Just `filter` instead? Or even `collect` to get the type cast directly instead of doing it below?